astropy/io/fits/tests/test_connect.py

```diff
@@ -842,6 +842,7 @@ def test_convert_comment_convention():
 
 
 def assert_objects_equal(obj1, obj2, attrs, compare_class=True):
+    __tracebackhide__ = True
```

Comment:
Not sure to understand the benefit. If a test fail I will want to know why and hence dive into this function to find which attribute and examine the value, so why hiding it ?

Reply:
I tried to explain the expected gain in #16080
I could furthermore add that:
- in a sense, this satisfies the principle of least surprise (as a user of `numpy.testing.assert_*` functions, I grew to expect the internals of such functions to be hidden in tracebacks)
- you can still opt-in the current behaviour using `--full-trace`, while the behaviour this is enabling is *not* achievable currently

Reply:
I personally would be more surprised by not being able to know why I had a failure. The traceback would still go through the failure within the unit test. Often I end up having to bisect issues by examining these helper functions and making temporary modifications to them.

If we do go through with this change I suggest we carefully consider the error messages being raised so that they are more indicative of what has caused the failure.

Reply:
Yes but being useful for `numpy.testing.assert_` doesn't mean it's always better. Here the code is specific to what is being tested and used by a few tests only, and there is a good chance that I will want to see the full traceback is there a problem.
Also the problem is that I don't know what to expect with `__tracebackhide__` activated, what will pytest print ?

Reply:
As you'll have seen from my comment above, I agree with @saimn here - I can see representations as being a single object where one wants to hide the traceback, since their content is uniform, but not a more complicated object where one also compares attributes.

Reply:
> If we do go through with this change I suggest we carefully consider the error messages being raised so that they are more indicative of what has caused the failure.

> Also the problem is that I don't know what to expect with __tracebackhide__ activated, what will pytest print ?

In short: pytest prints the exact same message as it would without `__tracebackhide__`, the difference is in the line that is associated with that message.

To illustrate let's take my `t.py` test module from #16071 (copied here in a dropdown for convenience)
<details><summary>t.py</summary>

```python
# t.py
import numpy as np

def assert_equal_len_v1(x, y):
    assert len(x) == len(y)

def assert_equal_len_v2(x, y):
    __tracebackhide__ = True
    assert len(x) == len(y)

def test_v1():
    a = np.arange(2)
    b = np.arange(3)
    assert_equal_len_v1(a, b)

def test_v2():
    a = np.arange(2)
    b = np.arange(3)
    assert_equal_len_v2(a, b)
```

</details>

Here's the output
```
$ pytest t.py
====================== test session starts ======================
platform darwin -- Python 3.12.0, pytest-8.0.0, pluggy-1.4.0
...
t.py FF                                                   [100%]

=========================== FAILURES ============================
____________________________ test_v1 ____________________________

    def test_v1():
        a = np.arange(2)
        b = np.arange(3)
>       assert_equal_len_v1(a, b)

t.py:14:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

x = array([0, 1]), y = array([0, 1, 2])

    def assert_equal_len_v1(x, y):
>       assert len(x) == len(y)
E       assert 2 == 3
E        +  where 2 = len(array([0, 1]))
E        +  and   3 = len(array([0, 1, 2]))

t.py:5: AssertionError
____________________________ test_v2 ____________________________

    def test_v2():
        a = np.arange(2)
        b = np.arange(3)
>       assert_equal_len_v2(a, b)
E       assert 2 == 3
E        +  where 2 = len(array([0, 1]))
E        +  and   3 = len(array([0, 1, 2]))

t.py:19: AssertionError
==================== short test summary info ====================
FAILED t.py::test_v1 - assert 2 == 3
FAILED t.py::test_v2 - assert 2 == 3
======================= 2 failed in 0.06s =======================
```